crates/core/src/metadata_store/providers/objstore/object_store_version_repository.rs

```diff
@@ -0,0 +1,442 @@
+
+        let tag = store.create(KEY_1, HELLO_WORLD).await.unwrap();
+
+        let tagged_value = store.get(KEY_1).await.unwrap();
+
+        assert_eq!(tagged_value.tag, tag);
+        assert_eq!(tagged_value.bytes, HELLO_WORLD);
+    }
+
+    #[tokio::test]
+    async fn get_non_existing_should_fail() {
+        let store = ObjectStoreVersionRepository::new_for_testing();
+
+        match store.get(KEY_1).await {
+            Err(VersionRepositoryError::NotFound) => {
+                // ok!
+            }
+            _ => {
+                panic!("Should be NotFound");
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn create_twice_should_fail() {
+        let store = ObjectStoreVersionRepository::new_for_testing();
+
+        store.create(KEY_1, HELLO_WORLD).await.unwrap();
+
+        match store.create(KEY_1, HELLO_WORLD).await {
+            Err(VersionRepositoryError::AlreadyExists) => {
+                // ok!
+            }
+            _ => {
+                panic!("should have failed");
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn delete_should_work() {
+        let store = ObjectStoreVersionRepository::new_for_testing();
+
+        store.create(KEY_1, HELLO_WORLD).await.unwrap();
+
+        store.delete(KEY_1).await.unwrap();
+
+        match store.get(KEY_1).await {
+            Err(VersionRepositoryError::NotFound) => {
+                // ok!
+            }
+            _ => {
+                panic!("should not be present");
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn create_after_delete_should_work() {
+        let store = ObjectStoreVersionRepository::new_for_testing();
+
+        store.create(KEY_1, HELLO_WORLD).await.unwrap();
+
+        store.delete(KEY_1).await.unwrap();
+
+        store.create(KEY_1, WORLD).await.unwrap();
+
+        let tv = store.get(KEY_1).await.unwrap();
+
+        assert_eq!(tv.bytes, WORLD);
+    }
+
+    #[tokio::test]
+    async fn conditional_put_should_work() {
+        let store = ObjectStoreVersionRepository::new_for_testing();
+
+        let tag = store.create(KEY_1, HELLO_WORLD).await.unwrap();
+
+        store.put_if_tag_matches(KEY_1, tag, WORLD).await.unwrap();
+
+        let tv = store.get(KEY_1).await.unwrap();
+
+        assert_eq!(tv.bytes, WORLD);
+    }
+
+    #[tokio::test]
+    async fn conditional_put_should_fail() {
+        let store = ObjectStoreVersionRepository::new_for_testing();
+
+        let tag = store.create(KEY_1, HELLO_WORLD).await.unwrap();
+
+        store
+            .put_if_tag_matches(KEY_1, tag.clone(), HELLO)
+            .await
+            .unwrap();
+
+        match store.put_if_tag_matches(KEY_1, tag.clone(), WORLD).await {
+            Err(VersionRepositoryError::PreconditionFailed) => {
+                // ok!
+            }
+            _ => panic!("should have failed"),
+        }
+    }
+
+    #[tokio::test]
+    async fn concurrency_test() {
```

Comment:
We could also spawn a multi threaded runtime via `tokio::test(flavor = "multi-thread")`. That way we also test parallelism. This is probably not super important, though.